Document KCC upgrade path (#23)
TBD54566975/known-customer-credential: README.md

2. PII is subsequently sent to the IDV vendor via the backend system for verification

> [!IMPORTANT]
> The 2 styles of integration provide flexibility for PFI IDV systems, but they don't impact Mobile Wallets at all, as

> > they have the same external behaviour. 

Reviewer: good callout!

This implementation involves 3 distinct participants that have different responsibilities:

### PFI

> The PFI operates a KCC issuer, as well as a tbDEX exchange.

Reviewer: interesting i've never thought to consider each PFI acting as a _tbdex exchange_. kinda cool.
other thoughts would be: "as a liquidity provider"

Author: mmm. I was struggling to describe what it is to run the tbDEX part of PFI. Happy to use "liquidity provider" as it'll align with other stuff

> The PFI operates a KCC issuer, as well as a tbDEX exchange.

Reviewer: do we feel like making explicit mention to `tbDEX` and `PFI`s is necessary vs. just `Issuer`? 

i know we mention of KCC issuance being an _extension_ of the work being done for tbDEX here:

![image](https://github.com/TBD54566975/known-customer-credential/assets/4887440/af3af945-7243-4d9a-a129-b7eb9676c33f)

thinking using "Issuer" would make it more generalized. wdyt @tomdaffurn ?

Author: Was unsure about explicit links to tbDEX. I added one for `presentation_definition_id` cos I think a reader will not know where that comes from. Same for this paragraph - goal is to explain how a wallet might find the PD ID

Reviewer: @tomdaffurn    I think we may be able to lean on on the fact that PDs in and of themselves are an established concept especially given that they're referenced in [presentation exchange](https://identity.foundation/presentation-exchange/spec/v2.0.0/#presentation-definition) [credential manifest](https://identity.foundation/credential-manifest/#credential-manifest) and [OID4VP](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#name-presentation_definition-par). ID is a required property of PDs. 

We can always mention how this works with tbDEX as an _example_. Vs. first classing tbdex and PFI as core concepts.

This implementation involves 3 distinct participants that have different responsibilities:

### PFI

> The PFI operates a KCC issuer, as well as a liquidity provider via tbDEX.

Reviewer 2: Oh I see @mistermoe beat me to this above

> * prevents Wallets / Mobile Apps from having to update source code in order to integrate with different PFIs
> * establish a clear distinction between the application initiating the flow and a PFI

## Multiple Credential Types

Financial service providers often perform KYC incrementally in response to customer's expressed intent to access services. For example, minimal upfront diligence on newly onboarded customers, followed by additional diligence when the customer's activity passes a threshold.

> A KCC Issuer may define multiple KCC types to represent different levels of KYC diligence, and describe each with a Presentation Definition. Mobile Apps who hold a KCC already may apply for new credentials to acquire a higher level KCC (see [Initiate Application](#request)) 

Reviewer 2: ```suggestion
A KCC Issuer may define multiple KCC types to represent different levels of KYC diligence, and describe each with a Presentation Definition. Mobile Apps which hold a previously issued KCC may apply for new credentials to acquire a higher level KCC (see [Initiate Application](#request)) 
```

# Credential Application Flow

```mermaid
sequenceDiagram

actor A as Applicant
participant W as Webview
participant D as Mobile App
participant P as PFI
participant I as IDV

rect rgba(0, 0, 0, 0.1)
    D->>+P: Initiate KCC application
    Note right of I: Initiate Application

>     P-->>-D: IDV Request

Reviewer 2: Nice love this swimlane

But why is this the only `-->>` (dotted) line?

Author: Using dotted lines for HTTP responses. I think the Issue KCC line should be dotted too

end
rect rgba(0, 0, 0, 0.1)
    D->>+W: Load URL
    Note right of I: Collect IDV
    A->>W: Provide identity data
    W->>+I: Applicant identity data
    deactivate W
end
rect rgba(0, 0, 0, 0.1)
    D->>+P: Request KCC
    Note right of I: Credential Issuance
    P->>-D: Issue KCC
end
```

## Initiate Application

> The application flow is initiated with a [SIOPv2](https://openid.github.io/SIOPv2/openid-connect-self-issued-v2-wg-draft.html) interaction that authenticates the applicant's DID.

Reviewer 2: 💯

#### Grant Type: urn:ietf:params:oauth:grant-type:pre-authorized_code
| Field                 | Description                                                                                              | Required | References                                                                                                                 | Comments |
| :-------------------- | :------------------------------------------------------------------------------------------------------- | :------- | :------------------------------------------------------------------------------------------------------------------------- | :------- |
| `pre-authorized_code` | The code representing the Credential Issuer's authorization for the Mobile App to obtain a credential | y        | [OID4VCI](https://openid.github.io/OpenID4VCI/openid-4-verifiable-credential-issuance-wg-draft.html#section-4.1.1-4.2.2.1) |          |

> [!WARNING] 
> TODO: explain rationale behind providing `credential_offer` at this stage

## IDV Collection

> The IDV Request sent to the Mobile App contains a `url` field. The Mobile App MUST load this URL in an embedded 

Reviewer 2: > The Mobile App MUST

This made me realize, we have `url` marked as required in the IDV Request table (above) but if the SIOPv2 Auth Response includes a compliant `vp_token` then the user won't go through IDV Collection, yeah?

Author: Fair question. Don't think the `vp_token` has been thought about deeply yet.

I think it would be possible to make `url` optional in a future spec version, but probably not possible to go the other way?

## Credential Issuance

> Credential Issuance is an [OID4VCI Pre-Authorized Code flow](https://openid.github.io/OpenID4VCI/openid-4-verifiable-credential-issuance-wg-draft.html#name-pre-authorized-code-flow). 

Reviewer 2: At some point we have to talk about the auth flow for repeat customers, right? I presume our KCC schema will require an expiration date. Perhaps this is fine for now.

Author: Mmhmm. Reissuing isn't clear to me. There was talk in Oakland of simply revoking any credential that becomes out of date, and issuing a wholly new one. Seem likely that we'll need user interaction for data refresh anyway. OID4VCI calls this out:

> Second, the Credential Issuer can reissue the Credential by starting the issuance process from the beginning. This would involve interaction with the End-User. A Credential needs to be reissued if the Wallet does not have a valid Access Token or a valid Refresh Token. With this approach, when a new Credential is issued, the Wallet might need to check if it already has a Credential of the same type and, if necessary, delete the old Credential. Otherwise, the Wallet might end up with more than one Credential of the same type, without knowing which one is the latest.

Reviewer 2: Should re-issuing always require users to resubmit PII?

Reviewer: In some cases yes in others no. For example, if we collected an mDL that expired we'd probably want them to represent
